tests/test_TFEngine.py

```diff
@@ -1096,6 +1099,80 @@ def test_attention_train_then_search():
   engine.finalize()
 
 
+def test_attention_subnetwork_base_dependency():
+    net_dict = {
+      "encoder": {
+        "class": "subnetwork",
+        "from": ["data:data"],
+        "subnetwork": {
+          "concat": {"class": "linear", "activation": "tanh", "n_out": 5},
+          "output": {'class': "copy", 'from': ['concat']}
+        }
+      },
+      "decoder": {
+        "class": "rec",
+        "from": [],
+        "target": "classes", "max_seq_len": 10,
+        "unit": {
+          'output': {'class': 'choice', 'target': 'classes', 'beam_size': 4, 'from': ["output_prob"]},
+          "end": {"class": "compare", "from": ["output"], "value": 0},
+          'orth_embed': {'class': 'linear', 'activation': None, 'from': ['output'], "n_out": 7},
+          "s": {"class": "rnn_cell", "unit": "LSTMBlock", "from": ["prev:att", "prev:orth_embed"], "n_out": 7},
+          "att": {
+            "class": "subnetwork",
+            "from": ["s", "prev:orth_embed"],
+            "concat_sources": False,
+            "subnetwork": {
+              "c_in": {"class": "linear", "activation": "tanh", "from": ["data:0", "data:1"], "n_out": 5},
+              "output": {"class": "dot_attention", "from": ["c_in"], "base": "base:base:encoder", "base_ctx": "base:base:encoder"},
+            }
+          },
+          "output_prob": {"class": "softmax", "from": ["prev:s", "att"], "target": "classes", "loss": "ce"}
+        },
+      },
+      "decision": {"class": "decide", "from": ["output"], "loss": "edit_distance"},
+      "output": {'class': "copy", 'from': ['decoder']}
+    }
+    run_dummy_training(net_dict)
+
+
+def test_attention_subnetwork_from_dependency():
+  net_dict = {
+    "encoder": {
+      "class": "subnetwork",
+      "from": ["data:data"],
+      "subnetwork": {
+        "concat": {"class": "linear", "activation": "tanh", "n_out": 5},
+        "output": {'class': "copy", 'from': ['concat']}
+      }
+    },
+    "decoder": {
+      "class": "rec",
+      "from": [],
+      "target": "classes", "max_seq_len": 10,
+      "unit": {
+        'output': {'class': 'choice', 'target': 'classes', 'beam_size': 4, 'from': ["output_prob"]},
+        "end": {"class": "compare", "from": ["output"], "value": 0},
+        'orth_embed': {'class': 'linear', 'activation': None, 'from': ['output'], "n_out": 7},
+        "s": {"class": "rnn_cell", "unit": "LSTMBlock", "from": ["prev:att", "prev:orth_embed"], "n_out": 7},
+        "att": {
+          "class": "subnetwork",
+          "from": ["s", "prev:orth_embed", "base:encoder"],
+          "concat_sources": False,
+          "subnetwork": {
+            "c_in": {"class": "linear", "activation": "tanh", "from": ["data:0", "data:1"], "n_out": 5},
```

Comment:
Btw, (not related to this PR, just generic comment), instead of `data:0` here, you could also simply directly use `base:s`. This is what I usually do. Isn't this cleaner?

Reply:
The best way is actually to define them in the "from" and then access it via the same name, so writing e.g. `data:s`. Because then you see exactly which external inputs the subnetwork has, but you also do not have to search for the index numbers.

Reply:
> access it via the same name, so writing e.g. `data:s`

This is actually the solution I dislike the most. This mixes up the data names and layer names. This could lead to strange cases when you have same data key names as layer names (unusual maybe, but so are all potential bugs in the beginning, until you get hit by them). And why is this better than just `base:s`? The `base:` prefix is clear what it does. What the `data:` prefix does is not so clear (you were searching long time for some bugs because of this, which would not have happened with `base:`).